ql/src/java/org/apache/hadoop/hive/ql/plan/mapper/PlanMapper.java

```diff
@@ -200,54 +199,46 @@ public void merge(Object o1, Object o2) {
   }
 
   private void link(Object o1, Object o2, boolean mayMerge) {
-
-    Set<Object> keySet = Collections.newSetFromMap(new IdentityHashMap<Object, Boolean>());
-    keySet.add(o1);
-    keySet.add(o2);
-    keySet.add(getKeyFor(o1));
-    keySet.add(getKeyFor(o2));
-
-    Set<EquivGroup> mGroups = Collections.newSetFromMap(new IdentityHashMap<EquivGroup, Boolean>());
-
-    for (Object object : keySet) {
-      EquivGroup group = objectMap.get(object);
-      if (group != null) {
-        mGroups.add(group);
-      }
+    // Caches signatures on the first access. A signature of an Operator could change as optimizers could mutate it,
+    // keeping its semantics. The current implementation caches the signature before optimizations so that we can
+    // link Operators with their signatures at consistent timing.
+    registerSignature(o1);
+    registerSignature(o2);
```

Comment:
> This change reverts that logic and takes signatures out of the equation for determining equivalent groups.

haven't read all of it - but if thats true - please don't do that...
looking at the volume of changes - it must be changing something fundamental...

what's the issue - is it an equiv  violation? can you put the stacktrace somewhere?
I think that should be fixed without altering this part - it might be a missing link; or a missing signature annotation somewhere - you should not change this at all... 

all these things could enable to back-map runtime statistict up to the calcite join planning phase...but I think that feature was never completed/enabled

Reply:
to get a failing test for these changes you would need 2 ops which have conflicting signatures stored into the metastore ; loaded back and they might get applied incorrectly...
possibly pretty hard to do...

Reply:
> Assuming that we do not allow implicit linking via signatures I don't know what's the point of registering the signature at this stage

That's because some optimization can mutate Operators, changing their true signatures. [One example I remember is TableScanPPD](https://github.com/apache/hive/blob/rel/release-4.0.0-beta-1/ql/src/java/org/apache/hadoop/hive/ql/ppd/OpProcFactory.java#L414). I'm not 100% sure which signatures should be used in that case, but it sounds more reasonable and predictable for me to use ones before optimization than ones in the middle of optimization. I would say Operators should not be mutable for the signing purpose, but it is a too fundamental change.

I will make a reply to @kgyrtkirk in another thread as I guess those comments were written before reading my gist.

Reply:
I missed mentioning what are affected by such optimizations. Many test cases on the following revision failed with `equivalence mapping violation`.
https://github.com/apache/hive/pull/5037/commits/217b26db7eee95a0d25725df985adc79705a80d6

That's because it tries to LINK a pre-optimized operator with a post-optimized operator.
1. Before optimization
    - Operator A has a signature SA and is grouped as GA
    - Operator B has a signature SB and is grouped as GB
2. AuxSignatureLinker doesn't MERGE GA and GB as A and B have different signatures. Note that non aux signatures are not cached here
3. After optimization
    - Operator B is optimized, and its signature becomes SA
4. StatsRulePsrocFactory tries to LINK GA and GB as A and B share the same signature SA -> equivalence mapping violation

Potentially, GA and GB are actually mergeable though the current implementation doesn't do so.